### PR TITLE
fix: Memory leak

### LINKED ISSUE #113: [BUG] Memory leak
The following is my code causing a memory leak in `ants@v2.4.2`
```
package main

import (
	"fmt"
	"net/http"
	_ "net/http/pprof"
	"sync"
	"time"

	"github.com/panjf2000/ants/v2"
)

func main() {
	/**
	pprof
	http://127.0.0.1:6060/debug/pprof/
	*/
	go func() {
		_ = http.ListenAndServe("localhost:6060", nil)
	}()

	for k := 0; k < 4; k++ {
		for i := 0; i < 400; i++ {
			go DoTask()
			time.Sleep(time.Millisecond * 10)
		}
		time.Sleep(time.Second * 2)
	}
	fmt.Println("Done")

}

func DoTask() {
	pool, _ := ants.NewPool(2, ants.WithExpiryDuration(time.Second*2))

	wg := sync.WaitGroup{}

	{
		wg.Add(1)
		err := pool.Submit(func() {
			defer wg.Done()
			time.Sleep(time.Millisecond * 10)
		})
		if err != nil {
			fmt.Println(err)
		}
	}

	{
		wg.Add(1)
		err := pool.Submit(func() {
			defer wg.Done()
			time.Sleep(time.Millisecond * 10)
		})
		if err != nil {
			fmt.Println(err)
		}
	}

	wg.Wait()
	pool.Release()
}

```

### Timeline
```
------------------------------------------------------------------------------------------
CPU0 revertWorker            -> state closed -> Lock -> reset -> Unlock
------------------------------------------------------------------------------------------
CPU1 Release     -> state ok ->                                         -> Lock -> insert
------------------------------------------------------------------------------------------
```

### Consider the following:
1. `Pool.revertWorker` starts to execute (`Pool.state` is not `CLOSED`), but `Pool.revertWorker` -> `Pool.lock.Lock` has not yet been called.
2. `Pool.Release` starts to execute (`Pool.state` is `CLOSED`) and `Pool.Release` -> `Pool.lock.Lock` has been called.
3. `Pool.purgePeriodically` would break out of a loop next time.
4. `Pool.Release` -> `p.workers.reset` (clean up all workers in `workerStack.items`).
5. `Pool.revertWorker` -> `p.workers.insert`(add a worker to `workerStack.items`), but this worker would never receive any task, also leading to `goWorker.run` leaks.

However, there may be a race condition between 1 and 2.

### Note:
I tried to solve that, it seemed to work.
```
 // revertWorker puts a worker back into free pool, recycling the goroutines.
 func (p *Pool) revertWorker(worker *goWorker) bool {
-       if capacity := p.Cap(); (capacity > 0 && p.Running() > capacity) || atomic.LoadInt32(&p.state) == CLOSED {
+       if capacity := p.Cap(); capacity > 0 && p.Running() > capacity {
                return false
        }
        worker.recycleTime = time.Now()
        p.lock.Lock()

+       if atomic.LoadInt32(&p.state) == CLOSED {
+               p.lock.Unlock()
+               return false
+       }
```

# In Chinese
我们发现`Pool.revertWorker`跟`Pool.Release`存在竞态问题，会导致`goroutine`泄露。

### Timeline
```
------------------------------------------------------------------------------------------
CPU0 revertWorker            -> state closed -> Lock -> reset -> Unlock
------------------------------------------------------------------------------------------
CPU1 Release     -> state ok ->                                         -> Lock -> insert
------------------------------------------------------------------------------------------
```

如果`Release`先运行（当前池子状态未关闭往下走等锁），然后`revertWorker`也开始运行了。只要后者比前者先拿到锁`Pool.lock`，就会导致这个worker泄露。
`Release`只会给当时存在池子中的`worker`发`nil`，而这个`worker`肯定没有在池子中；然后`purgePeriodically`也会因为池子关闭而退出循环，后面`worker`过期就没人管了。除此之外好像也没有其他方法可以让泄露的`worker`退出了。

### Note:
小弟斗胆给大佬您提交一个issue，说的不对的地方还请多多指正，蟹蟹！
想跟您更加深入的讨论一下这个库，方便加下您的微信吗？小弟邮箱cainiao666999@gmail.com，可以的话我来加您😃。

### PR DESCRIPTION
**Pull request by Mutated1994**
Fixes #113

### CODE REVIEW

**Comment by Mutated1994** [comment]
#113

File: pool.go [hunk 1]
```diff
@@ -253,12 +253,17 @@ func (p *Pool) retrieveWorker() (w *goWorker) {
 
 // revertWorker puts a worker back into free pool, recycling the goroutines.
 func (p *Pool) revertWorker(worker *goWorker) bool {
-	if capacity := p.Cap(); (capacity > 0 && p.Running() > capacity) || atomic.LoadInt32(&p.state) == CLOSED {
+	if capacity := p.Cap(); capacity > 0 && p.Running() > capacity {
```
Comment (panjf2000) [inline pool.go:256]: You shouldn't remove `atomic.LoadInt32(&p.state) == CLOSED` in this precheck, this condition is still useful for common cases, just add a double check in the lock scope.

**Review by panjf2000 — CHANGES_REQUESTED** [review]

**Comment by panjf2000** [comment]
Also please resolve [pool_func.go#L275](https://github.com/panjf2000/ants/blob/master/pool_func.go#L275) as well.

File: pool.go (hunk 1 above)
Comment (Mutated1994) [reply pool.go:256]: Bro, I got it.

**Comment by Mutated1994** [comment]
> Also please resolve this one:
> https://github.com/panjf2000/ants/blob/ad86bfa6d264d8c8ff8e4d083150cf147d4ac1ff/pool_func.go#L274-L286
> 
> 
> as well.



> Also please resolve this one:
> https://github.com/panjf2000/ants/blob/ad86bfa6d264d8c8ff8e4d083150cf147d4ac1ff/pool_func.go#L274-L286
> 
> 
> as well.

Sure, no problem

**Comment by Mutated1994** [comment]
> Also please resolve this one:
> https://github.com/panjf2000/ants/blob/ad86bfa6d264d8c8ff8e4d083150cf147d4ac1ff/pool_func.go#L274-L286
> 
> 
> as well.

Sure, no problem

**Review by panjf2000 — CHANGES_REQUESTED** [review]
I believe that this bugfix is intuitive enough, no need to add unit tests, please remove it.

**Comment by Mutated1994** [comment]
> I believe that this bugfix is intuitive enough, no need to add unit tests, please remove it.

Gotcha😃

**Review by panjf2000 — APPROVED** [review]

**Comment by panjf2000** [comment]
Thanks~

**Comment by Mutated1994** [comment]
> Thanks~

Excuse me, when will you consider adding a new tag.

**Comment by Mutated1994** [comment]
> Thanks~

@panjf2000 
Bro, when will you consider adding a new tag？😃

**Comment by coticom** [comment]
I have a point that I don't quite understand: After the double check, isn't it still possible that the pool might be closed? So, what's the purpose of this double check?

**Comment by OliverChao** [comment]
@coticom If the pool has been closed, `revertWorker` will never insert an un-released worker because of double check.

**Comment by ioworker0** [comment]
> I have a point that I don't quite understand: After the double check, isn't it still possible that the pool might be closed? So, what's the purpose of this double check?

IIUC, the first one looks more like is a best-effort early check. But it's
unreliable without the lock, so we also require the second one.

IOW, I don't believe that the first one is necessary in common cases :)

**Pull request merged**